jupyterlab_server/listings_handler.py

```diff
@@ -0,0 +1,119 @@
+"""Tornado handlers for dynamic listings loading."""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import os
+import time
+import json
+import re
+
+from .server import FileFindHandler
+
+import requests
+from tornado import gen, ioloop
+
+LISTINGS_URL_SUFFIX='@jupyterlab/extensionmanager-extension/listings.json'
+
+
+class ListingsHandler(FileFindHandler):
+    """A file handler that returns file content from the listings directory."""
+
+    blacklist_uris = set()
+    whitelist_uris = set()
+    blacklist = []
+    whitelist = []
+    listings = ''
+    initialized = False
+
+    def initialize(self, path, default_filename=None,
+                   no_cache_paths=None, listings_url=None):
+
+        if type(self.config['LabServerApp']['listings_refresh_ms']) == int:
+            self.listings_refresh_ms = self.config['LabServerApp']['listings_refresh_ms']
+        else:
+            self.listings_refresh_ms = 1000 * 60
+
+        FileFindHandler.initialize(self, path,
+                                   default_filename=default_filename,
+                                   no_cache_paths=no_cache_paths)
+
+        if not ListingsHandler.initialized:
+
+            if type(self.config['LabServerApp']['blacklist_uris']) == str:
+                ListingsHandler.blacklist_uris = set(self.config['LabServerApp']['blacklist_uris'].split(','))
+            if type(self.config['LabServerApp']['whitelist_uris']) == str:
+                ListingsHandler.whitelist_uris = set(self.config['LabServerApp']['whitelist_uris'].split(','))
+
+            self.init_listings_uris(os.path.join(path, LISTINGS_URL_SUFFIX))
+
+            self.fetch_listings_uris()
+
+            ListingsHandler.initialized = True
+
+            self.pc = ioloop.PeriodicCallback(self.fetch_listings_uris, callback_time=self.listings_refresh_ms)
+            self.pc.start()
+
+
+    def init_listings_uris(self, abs_path):
+        with open(abs_path, 'rb') as fid:
+            data = fid.read().decode('utf-8')
+        listings = json.loads(data)
+        if len(ListingsHandler.blacklist_uris) == 0:
+            for b in listings['listings']['blacklist_uris']:
+                ListingsHandler.blacklist_uris.add(b)
+        if len(ListingsHandler.whitelist_uris) == 0:
+            for w in listings['listings']['whitelist_uris']:
+                ListingsHandler.whitelist_uris.add(w)
+
+
+    def fetch_listings_uris(self):
+        self.log.info('Fetching Blacklist from {}'.format(ListingsHandler.blacklist_uris))
+        blacklist = []
+        for ListingsHandler.blacklist_uri in ListingsHandler.blacklist_uris:
+            r = requests.get(ListingsHandler.blacklist_uri)
```

Comment:
discussed in call:
- be careful with blocking
- in enterprise external requests might be blocked without proxies (e.g. allow for arbitrary kwargs)

Reply:
Thx Tim, good call. From https://requests.readthedocs.io/en/latest/user/advanced/#proxies a single additional trait, e.g. LabServerApp.http_proxy=http://user:pass@10.10.1.10:3128 would be enough. Any reason to foresee arbitrary kwargs?

Reply:
what if my resource is protected by cookie-based auth? I might need to pass in cookies to the request.

Reply:
So we can define a trait of type dictionary and use ` requests.request(method, url, **kwargs)`. https://2.python-requests.org/en/v2.7.0/api/#requests.request

I think it better maps what you were thinking to?

Reply:
Added here https://github.com/jupyterlab/jupyterlab_server/pull/82/commits/0318022bfbda5dc00d2d56df21494f0494cf5d2e#diff-39080f855774d3c6602514c700244e1fR43

Resolving converstation